In [ ]:
import pandas as pd
import mod_sql

#모듈 수정을 위해 미리 적어두는 reload
import imp
imp.reload(mod_sql)


_dt = mod_sql.Database()

In [4]:
sql = """
        SELECT * FROM sales
        """
result = _dt.executeAll(sql)
df = pd.DataFrame(result)
df

Region                           Country  \
0               Sub-Saharan Africa                              Chad   
1                           Europe                            Latvia   
2     Middle East and North Africa                          Pakistan   
3               Sub-Saharan Africa  Democratic Republic of the Congo   
4                           Europe                    Czech Republic   
...                            ...                               ...   
9995                          Asia                              Laos   
9996                        Europe                     Liechtenstein   
9997            Sub-Saharan Africa  Democratic Republic of the Congo   
9998            Sub-Saharan Africa                      South Africa   
9999                          Asia                          Mongolia   

            Item Type Sales Channel Order Priority  Order Date   Order ID  \
0     Office Supplies        Online              L   1/27/2011  292494523   
1           Beverages        Online              C  12/28/2015  361825549   
2          Vegetables       Offline              C   1/13/2011  141515767   
3           Household        Online              C   9/11/2012  500364005   
4           Beverages        Online              C  10/27/2015  127481591   
...               ...           ...            ...         ...        ...   
9995        Beverages        Online              H   7/15/2014  199342048   
9996        Cosmetics        Online              C  10/27/2012  763044106   
9997       Vegetables       Offline              M   2/14/2013  848579967   
9998             Meat        Online              L   2/19/2017  298185956   
9999           Snacks       Offline              M   4/12/2016  824410903   

      Ship Date  Units Sold  Unit Price  Unit Cost  Total Revenue  Total Cost  \
0     2/12/2011        4484      651.21     524.96     2920025.64  2353920.64   
1     1/23/2016        1075       47.45      31.79       51008.75    34174.25   
2      2/1/2011        6515      154.06      90.93     1003700.90   592408.95   
3     10/6/2012        7683      668.27     502.54     5134318.41  3861014.82   
4     12/5/2015        3491       47.45      31.79      165647.95   110978.89   
...         ...         ...         ...        ...            ...         ...   
9995  7/31/2014        8597       47.45      31.79      407927.65   273298.63   
9996  11/1/2012         562      437.20     263.33      245706.40   147991.46   
9997  3/20/2013        2524      154.06      90.93      388847.44   229507.32   
9998  2/22/2017        8706      421.89     364.69     3672974.34  3174991.14   
9999  4/16/2016         361      152.58      97.44       55081.38    35175.84   

      Total Profit  
0        566105.00  
1         16834.50  
2        411291.95  
3       1273303.59  
4         54669.06  
...            ...  
9995     134629.02  
9996      97714.94  
9997     159340.12  
9998     497983.20  
9999      19905.54  

[10000 rows x 14 columns]

최대 최소 출력해보기 

In [3]:
sql_2 = """
        SELECT MIN(`Units Sold`) FROM sales
        """
        #`Units Sold` 대소문자 안맞춰도 출력은 된다. 
result_2 = _dt.executeAll(sql_2)
df = pd.DataFrame(result_2)
df

MIN(`Units Sold`)
0                  2

In [8]:
sql_2 = """
        SELECT MAX(`Units Sold`) FROM sales
        """
        #`Units Sold` 대소문자 안맞춰도 출력은 된다. 
result_2 = _dt.executeAll(sql_2)
df = pd.DataFrame(result_2)
df

MAX(`Units Sold`)
0              10000

문자열은 최소 최대가 어떻게 나올까?

a 가 최소 
z 가 최대로 문자열 순서대로 출력된다. 

In [7]:
sql_3 = """
        SELECT MAX(`Region`) FROM sales
        """
result_3 = _dt.executeAll(sql_3)
df = pd.DataFrame(result_3)
df

MAX(`Region`)
0  Sub-Saharan Africa

### sql내 내장함수로 문자열 출력하기  
간혹 필요할 때가 있다. 
왜 영어는 1씩 잡히는데 한글은 여러개씩 잡힐까? 
length는 어떤 의미일까? 

한글은 예측하기 어려워서 한글자에 1byte가 들어간다. 
length 길이 체크라기 보다 스트링의 byte를 체크한다. 

sql 에서 컬럼을 만들때 VARCAHR(바이트 수) 이렇게 적어야 하는데, 
VARCHAR(32) 로 했을때 한글이 들어가야 된다면 10개 정도 밖에 들어가지 못한다는 계산이 나온다. 

그래서 미리 바이트를 글자수의 바이트를 체크해야한다고 할 때 필요하다.  

In [10]:
sql_4 = """
        SELECT LENGTH("Hello")
        """
_dt.executeAll(sql_4)

[{'LENGTH("Hello")': 5}]

In [11]:
sql_4 = """
        SELECT LENGTH("유리")
        """
_dt.executeAll(sql_4)

[{'LENGTH("유리")': 6}]

char_ 만 덧붙이면 글자 수가 출력된다.  

In [14]:
sql_5 = """
        SELECT CHAR_LENGTH("유리")
        """
_dt.executeAll(sql_5)

[{'CHAR_LENGTH("유리")': 2}]

위의 item type의 글자수를 알고싶다면? 

In [15]:
sql_6 = """
        SELECT CHAR_LENGTH(`Item Type`) FROM sales
        """
result_6 = _dt.executeAll(sql_6)
df_6 = pd.DataFrame(result_6)
df_6


CHAR_LENGTH(`Item Type`)
0                           15
1                            9
2                           10
3                            9
4                            9
...                        ...
9995                         9
9996                         9
9997                        10
9998                         4
9999                         6

[10000 rows x 1 columns]

위처럼 한 행렬마다 각각의 계산을 해서 해로운 컬럼을 넣는 것, 
파이썬에서 apply 함수를 통해 만들 수 있었다. 

파이선에선 apply를 통한 여러번의 반복작업을 slq 내장함수로는 한 줄로 끝낼 수 있다. 

물론 무엇이 더 편하냐는 개인차다. 

컬럼의 범주를 보여주는, 즉 
중복된 데이터값을 없애주는 내장함수가 있다. 

In [65]:
sql_7 = """
            SELECT DISTINCT Region 
            FROM sales
            ORDER BY Region ASC
        """
result_7 = _dt.executeAll(sql_7)
df_7 = pd.DataFrame(result_7)
df_7

Region
0                               Asia
1              Australia and Oceania
2  Central America and the Caribbean
3                             Europe
4       Middle East and North Africa
5                      North America
6                 Sub-Saharan Africa

어던 언어를 쓰든 조건문과 반복문을 사용할 수 있다. 

판다스의 apply lamda 처럼 sql도 사용할 수 있다. 

In [35]:
sql_8 = """
            SELECT * FROM emp
        """
result_8 = _dt.executeAll(sql_8)
df_8 = pd.DataFrame(result_8)
df_8

EMPNO   ENAME        JOB   MGR    HIREDATE      SAL     COMM DEPTNO
0   7369   SMITH      CLERK  7902  1980-12-17   800.00     None     20
1   7499   ALLEN   SALESMAN  7698  1981-02-20  1600.00   300.00     30
2   7521    WARD   SALESMAN  7698  1981-02-22  1250.00   500.00     30
3   7566   JONES    MANAGER  7839  1981-04-02  2975.00     None     20
4   7654  MARTIN   SALESMAN  7698  1981-09-28  1250.00  1400.00     30
5   7698   BLAKE    MANAGER  7839  1981-05-01  2850.00     None     30
6   7782   CLARK    MANAGER  7839  1981-06-09  2450.00     None     10
7   7788   SCOTT    ANALYST  7566  1987-06-28  3000.00     None     20
8   7839    KING  PRESIDENT  None  1981-11-17  5000.00     None     10
9   7844  TURNER   SALESMAN  7698  1981-09-08  1500.00     0.00     30
10  7876   ADAMS      CLERK  7788  1987-07-13  1100.00     None     20
11  7900   JAMES      CLERK  7698  1981-12-03   950.00     None     30
12  7902    FORD    ANALYST  7566  1981-12-03  3000.00     None     20
13  7934  MILLER      CLERK  7782  1982-01-23  1300.00     None     10

SAL	 : 새일즈
COMM	: 성과급
DEPTNO :부서번호

In [27]:
sql_8 = """
            SELECT 
            IF(SAL > 2000, 'HIGH', 'LOW') AS SAL
            FROM emp
        """
result_8 = _dt.executeAll(sql_8)
df_8 = pd.DataFrame(result_8)
df_8

SAL
0    LOW
1    LOW
2    LOW
3   HIGH
4    LOW
5   HIGH
6   HIGH
7   HIGH
8   HIGH
9    LOW
10   LOW
11   LOW
12  HIGH
13   LOW

In [26]:
sql_8 = """
            SELECT 
            ENAME, 
            IF(SAL > 2000, 'HIGH', 'LOW') AS SAL
            FROM emp
        """
result_8 = _dt.executeAll(sql_8)
df_8 = pd.DataFrame(result_8)
df_8

ENAME   SAL
0    SMITH   LOW
1    ALLEN   LOW
2     WARD   LOW
3    JONES  HIGH
4   MARTIN   LOW
5    BLAKE  HIGH
6    CLARK  HIGH
7    SCOTT  HIGH
8     KING  HIGH
9   TURNER   LOW
10   ADAMS   LOW
11   JAMES   LOW
12    FORD  HIGH
13  MILLER   LOW

COMM 커미션 컬럼의 결측치 값을 넣어주려면? 

In [28]:
sql_9 = """
            SELECT 
            ENAME, 
            IF(COMM, "1", "0") AS commition
            FROM emp
        """
result_9 = _dt.executeAll(sql_9)
df_9 = pd.DataFrame(result_9)
df_9

ENAME commition
0    SMITH         0
1    ALLEN         1
2     WARD         1
3    JONES         0
4   MARTIN         1
5    BLAKE         0
6    CLARK         0
7    SCOTT         0
8     KING         0
9   TURNER         0
10   ADAMS         0
11   JAMES         0
12    FORD         0
13  MILLER         0

하지만 위처럼 하면, 값이 있는 것들도 전부 0, 1로 바뀌고 만다. 
그렇지 않기 위해서는 어떻게 해야할까? 

판다스에선 fillna 를 썼다. sql에선? IFNULL

IFNULL(COMM, 0) AS commition
만약 null이면, 기준은 comm 으로,  0을 넣어라. 

In [29]:
sql_9 = """
            SELECT 
            ENAME, 
            IFNULL(COMM, 0) AS commition
            FROM emp
        """
result_9 = _dt.executeAll(sql_9)
df_9 = pd.DataFrame(result_9)
df_9

ENAME commition
0    SMITH      0.00
1    ALLEN    300.00
2     WARD    500.00
3    JONES      0.00
4   MARTIN   1400.00
5    BLAKE      0.00
6    CLARK      0.00
7    SCOTT      0.00
8     KING      0.00
9   TURNER      0.00
10   ADAMS      0.00
11   JAMES      0.00
12    FORD      0.00
13  MILLER      0.00

null을 대체하지 말고 그냥 체크하기만 하려면? 
nall이면 1이고, nall이 아니면 0 이 출력된다. 

In [32]:
sql_10 = """
            SELECT 
            ENAME, 
            ISNULL(COMM) AS commition
            FROM emp
        """
result_10 = _dt.executeAll(sql_10)
df_10 = pd.DataFrame(result_10)
df_10

ENAME  commition
0    SMITH          1
1    ALLEN          0
2     WARD          0
3    JONES          1
4   MARTIN          0
5    BLAKE          1
6    CLARK          1
7    SCOTT          1
8     KING          1
9   TURNER          0
10   ADAMS          1
11   JAMES          1
12    FORD          1
13  MILLER          1

특정 문자를 바꾸는 REPLACE 
SELECT REPLACE ("원본 문자열", "바꾸기 전 문자", "바꿀 문자")

In [33]:
sql_11 = """
            SELECT REPLACE ("A B C D", "B", "b")
        """
_dt.executeAll(sql_11)


[{'REPLACE ("A B C D", "B", "b")': 'A b C D'}]

In [36]:
df_8.head()

EMPNO   ENAME       JOB   MGR    HIREDATE      SAL     COMM DEPTNO
0  7369   SMITH     CLERK  7902  1980-12-17   800.00     None     20
1  7499   ALLEN  SALESMAN  7698  1981-02-20  1600.00   300.00     30
2  7521    WARD  SALESMAN  7698  1981-02-22  1250.00   500.00     30
3  7566   JONES   MANAGER  7839  1981-04-02  2975.00     None     20
4  7654  MARTIN  SALESMAN  7698  1981-09-28  1250.00  1400.00     30

emp 의 job 의 salesman을 판매사원이라고 바꿔본다. 

In [39]:
sql_11 = """
            SELECT 
            ENAME,
            REPLACE (JOB, "SALESMAN", "판매사원")
            FROM emp
        """
result_11 = _dt.executeAll(sql_11)
df_11 = pd.DataFrame(result_11)
df_11


ENAME REPLACE (JOB, "SALESMAN", "판매사원")
0    SMITH                             CLERK
1    ALLEN                              판매사원
2     WARD                              판매사원
3    JONES                           MANAGER
4   MARTIN                              판매사원
5    BLAKE                           MANAGER
6    CLARK                           MANAGER
7    SCOTT                           ANALYST
8     KING                         PRESIDENT
9   TURNER                              판매사원
10   ADAMS                             CLERK
11   JAMES                             CLERK
12    FORD                           ANALYST
13  MILLER                             CLERK

사칙연산을 쓰는 방법 

sal 에 1년이 지나서 재계약을 해야한다, 20% 인상을 해준다고 할때, 
각각의 연봉은 얼마나 올라갈까? 
sql * 1.2 

In [46]:
sql_12 = """
            SELECT 
            ENAME AS '사원명', 
            SAL AS '연봉', 
            SAL*1.2 AS '인상 후 연봉'
            FROM emp
        """
result_12 = _dt.executeAll(sql_12)
df_12 = pd.DataFrame(result_12)
df_12

사원명       연봉   인상 후 연봉
0    SMITH   800.00   960.000
1    ALLEN  1600.00  1920.000
2     WARD  1250.00  1500.000
3    JONES  2975.00  3570.000
4   MARTIN  1250.00  1500.000
5    BLAKE  2850.00  3420.000
6    CLARK  2450.00  2940.000
7    SCOTT  3000.00  3600.000
8     KING  5000.00  6000.000
9   TURNER  1500.00  1800.000
10   ADAMS  1100.00  1320.000
11   JAMES   950.00  1140.000
12    FORD  3000.00  3600.000
13  MILLER  1300.00  1560.000

그룹을 만드는 법

부서 번호(DEPTNO)를 가지고 그룹바이 한 다음에 그룹에 있는 사람들을 컬럼하나의 리스트 형태로 뽑아보자 

In [48]:
sql_13 = """
            SELECT 
            DEPTNO 
            FROM emp 
            GROUP BY DEPTNO    
        """
result_13 = _dt.executeAll(sql_13)
df_13 = pd.DataFrame(result_13)
df_13

DEPTNO
0     10
1     20
2     30

이 부서에 속한 사람을 그대로 보고싶다면? 
select 로 출력하면 그냥 각 부서의 첫번째 사람만 나온다. 

In [50]:
sql_13 = """
            SELECT 
            DEPTNO, ENAME 
            FROM emp 
            GROUP BY DEPTNO    
        """
result_13 = _dt.executeAll(sql_13)
df_13 = pd.DataFrame(result_13)
df_13

DEPTNO  ENAME
0     10  CLARK
1     20  SMITH
2     30  ALLEN

GROUP_CONCAT(ENAME) 
그룹을 연결해 준다는 의미로 사용한다. 

In [52]:
sql_13 = """
            SELECT 
            DEPTNO,
            GROUP_CONCAT(ENAME)
            FROM emp 
            GROUP BY DEPTNO    
        """
result_13 = _dt.executeAll(sql_13)
df_13 = pd.DataFrame(result_13)
df_13

DEPTNO                   GROUP_CONCAT(ENAME)
0     10                     CLARK,KING,MILLER
1     20          SMITH,JONES,SCOTT,ADAMS,FORD
2     30  ALLEN,WARD,MARTIN,BLAKE,TURNER,JAMES

보고싶은 항목을 추가할 수 있다. 

In [54]:
sql_13 = """
            SELECT 
            DEPTNO,
            GROUP_CONCAT(ENAME),
            GROUP_CONCAT(JOB)
            FROM emp 
            GROUP BY DEPTNO    
        """
result_13 = _dt.executeAll(sql_13)
df_13 = pd.DataFrame(result_13)
df_13

DEPTNO                   GROUP_CONCAT(ENAME)  \
0     10                     CLARK,KING,MILLER   
1     20          SMITH,JONES,SCOTT,ADAMS,FORD   
2     30  ALLEN,WARD,MARTIN,BLAKE,TURNER,JAMES   

                                   GROUP_CONCAT(JOB)  
0                            MANAGER,PRESIDENT,CLERK  
1                CLERK,MANAGER,ANALYST,CLERK,ANALYST  
2  SALESMAN,SALESMAN,SALESMAN,MANAGER,SALESMAN,CLERK

CANCAT( ) ? 

데이터 베이스 컬럼이 너무 많아서 합치는 경우에 주로 사용한다. 

파이썬에서 sql 문을 쓸때 %s 와 같은 문자열을 잘 인식하기 어려워서 이렇게 concat을 많이 쓸 수 밖에 없다.  

In [55]:
sql_14 = """
            SELECT 
            CONCAT("a","b","c","d")
        """
_dt.executeAll(sql_14)

[{'CONCAT("a","b","c","d")': 'abcd'}]

In [56]:
sql_15 = """
            SELECT 
            CONCAT (ENAME, '(', JOB, ')') AS ENAME
            FROM emp
        """
result_15 = _dt.executeAll(sql_15)
df_15 = pd.DataFrame(result_15)
df_15

ENAME
0       SMITH(CLERK)
1    ALLEN(SALESMAN)
2     WARD(SALESMAN)
3     JONES(MANAGER)
4   MARTIN(SALESMAN)
5     BLAKE(MANAGER)
6     CLARK(MANAGER)
7     SCOTT(ANALYST)
8    KING(PRESIDENT)
9   TURNER(SALESMAN)
10      ADAMS(CLERK)
11      JAMES(CLERK)
12     FORD(ANALYST)
13     MILLER(CLERK)

repalce를 통해 하나만 바꿔줬었는데, 여러개 동시에 바꿔줄때는 case를 사용한다. 

In [60]:
sql_16 = """
            SELECT 
            ENAME,
            CASE
                WHEN JOB = "ANALYST" THEN "분석가"
                WHEN JOB = "CLERK" THEN "점원"
                WHEN JOB = "MANAGER" THEN "관리자"
                WHEN JOB = "PRESIDENT" THEN "사장" 
                WHEN JOB = "SALESMAN" THEN "영업사원"
                END AS JOB
            FROM emp 
        """
result_16 = _dt.executeAll(sql_16)
df_16 = pd.DataFrame(result_16)
df_16

ENAME   JOB
0    SMITH    점원
1    ALLEN  영업사원
2     WARD  영업사원
3    JONES   관리자
4   MARTIN  영업사원
5    BLAKE   관리자
6    CLARK   관리자
7    SCOTT   분석가
8     KING    사장
9   TURNER  영업사원
10   ADAMS    점원
11   JAMES    점원
12    FORD   분석가
13  MILLER    점원

if 랑 case의 차이는 
조건 하나 걸때는 if 
조건 여러개 걸 때는 case가 훨씬 낫다. 다중if 라고 생각하면 된다 if elif elif 처럼. 

그래서 else 도 쓸 수 있다. 

when 이 말하자면 if or elif고 
then 이 줄바꿈이라고 할 수 있다. 

In [61]:
sql_16 = """
            SELECT 
            ENAME,
            CASE
                WHEN JOB = "ANALYST" THEN "분석가"
                WHEN JOB = "CLERK" THEN "점원"
                WHEN JOB = "MANAGER" THEN "관리자"
                WHEN JOB = "PRESIDENT" THEN "사장" 
                ELSE "영업사원"
                END AS JOB
            FROM emp 
        """
result_16 = _dt.executeAll(sql_16)
df_16 = pd.DataFrame(result_16)
df_16

ENAME   JOB
0    SMITH    점원
1    ALLEN  영업사원
2     WARD  영업사원
3    JONES   관리자
4   MARTIN  영업사원
5    BLAKE   관리자
6    CLARK   관리자
7    SCOTT   분석가
8     KING    사장
9   TURNER  영업사원
10   ADAMS    점원
11   JAMES    점원
12    FORD   분석가
13  MILLER    점원

1. emp table 과 dept table 을 left join 조건 deptno가 같은 경우 
2. 출력 컬럼은 ENAME, JOB, SAL, DEPTNO, LOC
3. ENAME, LOC 그대로 출력 
4. JOB은 각각 한글명으로 변경 
5. SAL 은 2450을 기준으로 2450이상이면 HIGH 미만이면 LOW 
6. DEPTNO는 DEPTNO값과 DEPT table의 DANAME 값을 더해서 표시 (DEPTINO - DNAME형식)
7. 모든 칼럼의 이름은 한글로 변경 

In [81]:
sql_17 = """ 
            SELECT emp.ENAME, emp.JOB, emp.SAL, emp.DEPTNO, dept.LOC,
            CASE
                WHEN JOB = "ANALYST" THEN "분석가"
                WHEN JOB = "CLERK" THEN "점원"
                WHEN JOB = "MANAGER" THEN "관리자"
                WHEN JOB = "PRESIDENT" THEN "사장" 
                WHEn JOB =  "SALESMAN" THEN "영업사원"
            END AS JOB
            FROM emp
            LEFT JOIN dept
            ON emp.DEPTNO = dept.DEPTNO
        """
result_17 = _dt.executeAll(sql_17)
df = pd.DataFrame(result_17)
print(df)

     ENAME        JOB      SAL DEPTNO       LOC  .JOB
0    SMITH      CLERK   800.00     20    DALLAS    점원
1    ALLEN   SALESMAN  1600.00     30   CHICAGO  영업사원
2     WARD   SALESMAN  1250.00     30   CHICAGO  영업사원
3    JONES    MANAGER  2975.00     20    DALLAS   관리자
4   MARTIN   SALESMAN  1250.00     30   CHICAGO  영업사원
5    BLAKE    MANAGER  2850.00     30   CHICAGO   관리자
6    CLARK    MANAGER  2450.00     10  NEW YORK   관리자
7    SCOTT    ANALYST  3000.00     20    DALLAS   분석가
8     KING  PRESIDENT  5000.00     10  NEW YORK    사장
9   TURNER   SALESMAN  1500.00     30   CHICAGO  영업사원
10   ADAMS      CLERK  1100.00     20    DALLAS    점원
11   JAMES      CLERK   950.00     30   CHICAGO    점원
12    FORD    ANALYST  3000.00     20    DALLAS   분석가
13  MILLER      CLERK  1300.00     10  NEW YORK    점원


코드를 짤때는 단계별로 짜는 것이 좋다. 

무엇과 무엇의 LEFT JOIN 인지, 무엇을 기준으로 하는지 작성

In [ ]:
sql_17 = """
            SELECT 
        
            FROM emp LEFT JOIN dept 
            ON emp.DEPTNO = dept.DEPTNO
        """


출력할 컬럼 하나하나 작성 
ENAME, JOB, SAL, DEPTNO, LOC

In [ ]:
sql_17 = """
            SELECT 
            emp.ENAME,
            emp.JOB,
            emp.SAL,
            dept.LOC
            FROM emp LEFT JOIN dept 
            ON emp.DEPTNO = dept.DEPTNO
        """


출력하는 컬럼의 서브 조건들 작성

In [ ]:
sql_17 = """
            SELECT 
            emp.ENAME,
            dept.LOC,
            CASE
                WHEN emp.JOB = "ANALYST" THEN "분석가"
                WHEN emp.JOB = "CLERK" THEN "점원"
                WHEN emp.JOB = "MANAGER" THEN "관리자"
                WHEN emp.JOB = "PRESIDENT" THEN "사장" 
                WHEn emp.JOB =  "SALESMAN" THEN "영업사원"
            END,
            emp.SAL,
            dept.LOC
            FROM emp LEFT JOIN dept 
            ON emp.DEPTNO = dept.DEPTNO
        """

In [89]:
sql_17 = """
            SELECT 
            emp.ENAME,
            dept.LOC,
            CASE
                WHEN emp.JOB = "ANALYST" THEN "분석가"
                WHEN emp.JOB = "CLERK" THEN "점원"
                WHEN emp.JOB = "MANAGER" THEN "관리자"
                WHEN emp.JOB = "PRESIDENT" THEN "사장" 
                WHEn emp.JOB =  "SALESMAN" THEN "영업사원"
            END,
            IF (emp.SAL >= 2450, "HIGH", "LOW"),
            CONCAT(emp.DEPTNO, "-", dept.DNAME)
            FROM emp LEFT JOIN dept 
            ON emp.DEPTNO = dept.DEPTNO
        """
result_17 = _dt.executeAll(sql_17)
pd.DataFrame(result_17)

ENAME       LOC  \
0    SMITH    DALLAS   
1    ALLEN   CHICAGO   
2     WARD   CHICAGO   
3    JONES    DALLAS   
4   MARTIN   CHICAGO   
5    BLAKE   CHICAGO   
6    CLARK  NEW YORK   
7    SCOTT    DALLAS   
8     KING  NEW YORK   
9   TURNER   CHICAGO   
10   ADAMS    DALLAS   
11   JAMES   CHICAGO   
12    FORD    DALLAS   
13  MILLER  NEW YORK   

   CASE\n                WHEN emp.JOB = "ANALYST" THEN "분석가"\n                WHEN emp.JOB = "CLERK" THEN "점원"\n                WHEN emp.JOB = "MANAGER" THEN "관리자"\n                WHEN emp.JOB = "PRESIDENT" THEN "사장" \n                WHEn emp  \
0                                                  점원                                                                                                                                                                                                  
1                                                영업사원                                                                                                                                                                                                  
2                                                영업사원                                                                                                                                                                                                  
3                                                 관리자                                                                                                                                                                                                  
4                                                영업사원                                                                                                                                                                                                  
5                                                 관리자                                                                                                                                                                                                  
6                                                 관리자                                                                                                                                                                                                  
7                                                 분석가                                                                                                                                                                                                  
8                                                  사장                                                                                                                                                                                                  
9                                                영업사원                                                                                                                                                                                                  
10                                                 점원                                                                                                                                                                                                  
11                                                 점원                                                                                                                                                                                                  
12                                                분석가                                                                                                                                                                                                  
13                                                 점원                                                                                                                       

이름까지 변경 

In [94]:
sql_17 = """
            SELECT 
            emp.ENAME AS "사원명",
            dept.LOC AS "위치",
            CASE
                WHEN emp.JOB = "ANALYST" THEN "분석가"
                WHEN emp.JOB = "CLERK" THEN "점원"
                WHEN emp.JOB = "MANAGER" THEN "관리자"
                WHEN emp.JOB = "PRESIDENT" THEN "사장" 
                WHEn emp.JOB =  "SALESMAN" THEN "영업사원"
            END AS "직급",
            IF (emp.SAL >= 2450, "HIGH", "LOW") AS "월급",
            CONCAT(emp.DEPTNO, "-", dept.DNAME) AS "부서번호-부서명"
            FROM emp LEFT JOIN dept 
            ON emp.DEPTNO = dept.DEPTNO
            ORDER BY 사원명 ASC
        """
result_17 = _dt.executeAll(sql_17)
pd.DataFrame(result_17)


사원명        위치    직급    월급       부서번호-부서명
0    ADAMS    DALLAS    점원   LOW    20-RESEARCH
1    ALLEN   CHICAGO  영업사원   LOW       30-SALES
2    BLAKE   CHICAGO   관리자  HIGH       30-SALES
3    CLARK  NEW YORK   관리자  HIGH  10-ACCOUNTING
4     FORD    DALLAS   분석가  HIGH    20-RESEARCH
5    JAMES   CHICAGO    점원   LOW       30-SALES
6    JONES    DALLAS   관리자  HIGH    20-RESEARCH
7     KING  NEW YORK    사장  HIGH  10-ACCOUNTING
8   MARTIN   CHICAGO  영업사원   LOW       30-SALES
9   MILLER  NEW YORK    점원   LOW  10-ACCOUNTING
10   SCOTT    DALLAS   분석가  HIGH    20-RESEARCH
11   SMITH    DALLAS    점원   LOW    20-RESEARCH
12  TURNER   CHICAGO  영업사원   LOW       30-SALES
13    WARD   CHICAGO  영업사원   LOW       30-SALES

컬럼의 이름이 중첩 되는 애들은 어느테이블에 있는 컬럼 값인지 정확히 정해줘야하는데 
중첩되지 않는, 즉 두 테이블 중 하나에만 존재하는 컬럼명은 소속을 적어주지 않아도 출력된다. 
ex, emp.DEPTNO 는 dept.DEPTNO 가 있으니 구분 필요 
JOB 은 emp만 있으니 적지 않아도 출력됨. 

SELECT 하기 전에 from을 보고 출력하려고 하니 거기에 left join이 ㅣㅇㅆ어 
근데 그 기준이 emp.deptno == deptno래 
근데 그 기준이 ORDER BY `사원명` 이래 

이걸 다 하고 SELECT 한 거임 

즉, 일반적인 위에서 아래로 내려오는 프로그래밍 과는 다르다. 

타인 서버에 한거번에 로그인 해보기 

In [96]:
sql_18 = """
            SELECT * FROM emp
            WHERE DEPTNO > 10 
        """
result_18 = _dt.executeAll(sql_18)
pd.DataFrame(result_18)

EMPNO   ENAME       JOB   MGR    HIREDATE      SAL     COMM DEPTNO
0   7369   SMITH     CLERK  7902  1980-12-17   800.00     None     20
1   7499   ALLEN  SALESMAN  7698  1981-02-20  1600.00   300.00     30
2   7521    WARD  SALESMAN  7698  1981-02-22  1250.00   500.00     30
3   7566   JONES   MANAGER  7839  1981-04-02  2975.00     None     20
4   7654  MARTIN  SALESMAN  7698  1981-09-28  1250.00  1400.00     30
5   7698   BLAKE   MANAGER  7839  1981-05-01  2850.00     None     30
6   7788   SCOTT   ANALYST  7566  1987-06-28  3000.00     None     20
7   7844  TURNER  SALESMAN  7698  1981-09-08  1500.00     0.00     30
8   7876   ADAMS     CLERK  7788  1987-07-13  1100.00     None     20
9   7900   JAMES     CLERK  7698  1981-12-03   950.00     None     30
10  7902    FORD   ANALYST  7566  1981-12-03  3000.00     None     20

IN ( )  을 넣으면? 
WHERE 조건이 많을수록 in(  ) 을 사용하면 유리하다 

In [99]:
sql_18 = """
            SELECT * FROM emp
            WHERE DEPTNO IN (10) 
        """
result_18 = _dt.executeAll(sql_18)
pd.DataFrame(result_18)

EMPNO   ENAME        JOB   MGR    HIREDATE      SAL  COMM DEPTNO
0  7782   CLARK    MANAGER  7839  1981-06-09  2450.00  None     10
1  7839    KING  PRESIDENT  None  1981-11-17  5000.00  None     10
2  7934  MILLER      CLERK  7782  1982-01-23  1300.00  None     10

두개도 가능

In [100]:
sql_18 = """
            SELECT * FROM emp
            WHERE DEPTNO IN (10, 30) 
        """
result_18 = _dt.executeAll(sql_18)
pd.DataFrame(result_18)

EMPNO   ENAME        JOB   MGR    HIREDATE      SAL     COMM DEPTNO
0  7782   CLARK    MANAGER  7839  1981-06-09  2450.00     None     10
1  7839    KING  PRESIDENT  None  1981-11-17  5000.00     None     10
2  7934  MILLER      CLERK  7782  1982-01-23  1300.00     None     10
3  7499   ALLEN   SALESMAN  7698  1981-02-20  1600.00   300.00     30
4  7521    WARD   SALESMAN  7698  1981-02-22  1250.00   500.00     30
5  7654  MARTIN   SALESMAN  7698  1981-09-28  1250.00  1400.00     30
6  7698   BLAKE    MANAGER  7839  1981-05-01  2850.00     None     30
7  7844  TURNER   SALESMAN  7698  1981-09-08  1500.00     0.00     30
8  7900   JAMES      CLERK  7698  1981-12-03   950.00     None     30

같은 구문을 WHERE에서 줬어야 했다면? 

In [102]:
sql_18 = """
            SELECT * FROM emp
            WHERE 
            DEPTNO = 10 OR DEPTNO = 30
        """
result_18 = _dt.executeAll(sql_18)
pd.DataFrame(result_18)

EMPNO   ENAME        JOB   MGR    HIREDATE      SAL     COMM DEPTNO
0  7782   CLARK    MANAGER  7839  1981-06-09  2450.00     None     10
1  7839    KING  PRESIDENT  None  1981-11-17  5000.00     None     10
2  7934  MILLER      CLERK  7782  1982-01-23  1300.00     None     10
3  7499   ALLEN   SALESMAN  7698  1981-02-20  1600.00   300.00     30
4  7521    WARD   SALESMAN  7698  1981-02-22  1250.00   500.00     30
5  7654  MARTIN   SALESMAN  7698  1981-09-28  1250.00  1400.00     30
6  7698   BLAKE    MANAGER  7839  1981-05-01  2850.00     None     30
7  7844  TURNER   SALESMAN  7698  1981-09-08  1500.00     0.00     30
8  7900   JAMES      CLERK  7698  1981-12-03   950.00     None     30

아닌 값을 출력하는 방법 

In [103]:
sql_18 = """
            SELECT * FROM emp
            WHERE DEPTNO NOT IN (10, 30) 
        """
result_18 = _dt.executeAll(sql_18)
pd.DataFrame(result_18)

EMPNO  ENAME      JOB   MGR    HIREDATE      SAL  COMM DEPTNO
0  7369  SMITH    CLERK  7902  1980-12-17   800.00  None     20
1  7566  JONES  MANAGER  7839  1981-04-02  2975.00  None     20
2  7788  SCOTT  ANALYST  7566  1987-06-28  3000.00  None     20
3  7876  ADAMS    CLERK  7788  1987-07-13  1100.00  None     20
4  7902   FORD  ANALYST  7566  1981-12-03  3000.00  None     20

( ) 을 사용하는 것은  ( ) 안에 SQL 문을 넣기 위함이다. 

In [106]:
sql_19 = """
            SELECT
            (
                SELECT COUNT(*) FROM emp 
            ) AS `cnt(emp)`,
            (
                SELECT COUNT(*) FROM dept
            )AS `cnt(dept)`
        """
result_19 = _dt.executeAll(sql_19)
pd.DataFrame(result_19)

cnt(emp)  cnt(dept)
0        14          4

### emp table
>사원정보를 담고 있는 테이블

### dept table
>부서의 정보를 담고 있는 테이블 

### 부서의 이름이 'sales'인 소속 사원들의 리스트를 출력하라 

<실제 SQL 질문이 이럼>


emp에서 name 값을 출력해야하는데 
부서 이름이 sales인 사람만 출력해야한다. 


먼저 다중쿼리 안쓰고 출력하려면? 

In [111]:
sql_20 = """
        SELECT 
        ENAME
        FROM emp
        LEFT JOIN dept
        ON emp.DEPTNO = dept.DEPTNO
        WHERE dept.DNAME = "sales"  
        """
result_20 = _dt.executeAll(sql_20)
pd.DataFrame(result_20)

ENAME
0   ALLEN
1    WARD
2  MARTIN
3   BLAKE
4  TURNER
5   JAMES

다중쿼리 하려면 
slaes라고 하는 부서명을 가진 dept와 
emp 부서명이 같은 것을 출력하면 된ㄷ. 

In [113]:
sql_21 = """
        SELECT 
        ENAME
        FROM emp
        WHERE DEPTNO =  
        (
                SELECT DEPTNO FROM dept
                WHERE DNAME = "sales"
        )
        """
result_21 = _dt.executeAll(sql_21)
pd.DataFrame(result_21)

ENAME
0   ALLEN
1    WARD
2  MARTIN
3   BLAKE
4  TURNER
5   JAMES

부서의 위치가 'NEW YORK', 'DALLAS'가 아닌 소속 사원들의 리스트를 출력하라 

부서 위치를 알기 위해 부서번호를 통해 비교해야한다. 

In [125]:
sql_22 = """
            SELECT ENAME
            FROM emp
            WHERE 
            DEPTNO = 
            (
                SELECT 
                DEPTNO
                FROM dept 
                WHERE 
                LOC NOT IN ('NEW YORK','DALLAS')
            )
        """
result_22 = _dt.executeAll(sql_22)
pd.DataFrame(result_22)        

OperationalError: (1242, 'Subquery returns more than 1 row')

1242 에러가 나는게 정상이다. 왜나면 DEPTNO라는 부서 정보가 많다. 뉴옥과 달라스가 아닌 데이터는 시카고와 다른 지역이 있는데. 

값이 2개인, 즉 하나가 아닌 경우는 = 보단 IN으로 바꿔주면 된다. 


In [126]:
sql_22 = """
            SELECT ENAME
            FROM emp
            WHERE 
            DEPTNO IN
            (
                SELECT 
                DEPTNO
                FROM dept 
                WHERE 
                LOC NOT IN ('NEW YORK','DALLAS')
            )
        """
result_22 = _dt.executeAll(sql_22)
pd.DataFrame(result_22)    

ENAME
0   ALLEN
1    WARD
2  MARTIN
3   BLAKE
4  TURNER
5   JAMES

             SELECT 
                DEPTNO
                FROM dept 
                WHERE 
                LOC NOT IN ('NEW YORK','DALLAS') 
                

            를 쳐보면 값이 한개가 아니다. 두개는 나온다. 
            그래서 한개 값을 비교하는 = 이 아니라 
            in을 써야 한다. 

WHERE 를 쓰는 조건 중 하나로 LIKE를 써보자 
월급 ~보다 크고  ~~보다 작다 

In [128]:
sql_23 = """
            SELECT * FROM emp 
            WHERE SAL >= 2000 AND SAL < 5000
        """
result_23 = _dt.executeAll(sql_23)
pd.DataFrame(result_23)

EMPNO  ENAME      JOB   MGR    HIREDATE      SAL  COMM DEPTNO
0  7566  JONES  MANAGER  7839  1981-04-02  2975.00  None     20
1  7698  BLAKE  MANAGER  7839  1981-05-01  2850.00  None     30
2  7782  CLARK  MANAGER  7839  1981-06-09  2450.00  None     10
3  7788  SCOTT  ANALYST  7566  1987-06-28  3000.00  None     20
4  7902   FORD  ANALYST  7566  1981-12-03  3000.00  None     20

In [130]:
sql_24 = """
            SELECT * FROM emp 
            WHERE SAL BETWEEN 2000 AND 5000
        """
result_24 = _dt.executeAll(sql_24)
pd.DataFrame(result_24)

EMPNO  ENAME        JOB   MGR    HIREDATE      SAL  COMM DEPTNO
0  7566  JONES    MANAGER  7839  1981-04-02  2975.00  None     20
1  7698  BLAKE    MANAGER  7839  1981-05-01  2850.00  None     30
2  7782  CLARK    MANAGER  7839  1981-06-09  2450.00  None     10
3  7788  SCOTT    ANALYST  7566  1987-06-28  3000.00  None     20
4  7839   KING  PRESIDENT  None  1981-11-17  5000.00  None     10
5  7902   FORD    ANALYST  7566  1981-12-03  3000.00  None     20

btween은 두 값이 포함된 것이 나온다 

만약에 s로 시작하는 이름의 리스트를 출력하시오 라고 한다면? 
like 없이 어떻게 할까? 

ENAME 문자열을 받아와서 Split ( ) 한 뒤 , 하나씩 뽑아서 추출해서 S로 시작하면 출력하고, 아니면 넘어가도록 IF문을짜면 될텐데. 

이런것을 단번에 푸는 방법이 LIKE다 


워크 벤치에서 쓸때는 

SELECT * FROM ubion.emp WHERE ENAME LIKE 'S%';

SELECT * FROM ubion.emp WHERE ENAME LIKE '%S%'; 는 앞뒤 상관없이 그냥 S가 들어가기만 하면 다 출력한다. 

SELECT * FROM ubion.emp WHERE ENAME LIKE '%S';를 쓰면 맨 마지막에 S가 붙은 것만 출력된다. 

하지만 이 코드를 파이썬 안에서 쓸 수는 없다. 

In [150]:
sql_25 = """

            SELECT * FROM emp WHERE ENAME LIKE '%S'
        """
result_24 = _dt.executeAll(sql_25)
pd.DataFrame(result_24)

ValueError: unsupported format character 'S' (0x53) at index 51

파이썬에서 %s , %d, %f 로 리스트에 있는 값을 담아주는 문구다. 
그래서 SQL문을 사용할 수 없다. 
이를 해결하기 위해서는 %%를 사용한다. 

In [133]:
sql_25 = """

            SELECT * FROM emp WHERE ENAME LIKE 'S%%'
        """
result_24 = _dt.executeAll(sql_25)
pd.DataFrame(result_24)

EMPNO  ENAME      JOB   MGR    HIREDATE      SAL  COMM DEPTNO
0  7369  SMITH    CLERK  7902  1980-12-17   800.00  None     20
1  7788  SCOTT  ANALYST  7566  1987-06-28  3000.00  None     20

In [134]:
sql_25 = """

            SELECT * FROM emp WHERE ENAME LIKE '%%S'
        """
result_24 = _dt.executeAll(sql_25)
pd.DataFrame(result_24)

EMPNO  ENAME      JOB   MGR    HIREDATE      SAL  COMM DEPTNO
0  7566  JONES  MANAGER  7839  1981-04-02  2975.00  None     20
1  7876  ADAMS    CLERK  7788  1987-07-13  1100.00  None     20
2  7900  JAMES    CLERK  7698  1981-12-03   950.00  None     30

In [135]:
sql_25 = """

            SELECT * FROM emp WHERE ENAME LIKE '%%S%%'
        """
result_24 = _dt.executeAll(sql_25)
pd.DataFrame(result_24)

EMPNO  ENAME      JOB   MGR    HIREDATE      SAL  COMM DEPTNO
0  7369  SMITH    CLERK  7902  1980-12-17   800.00  None     20
1  7566  JONES  MANAGER  7839  1981-04-02  2975.00  None     20
2  7788  SCOTT  ANALYST  7566  1987-06-28  3000.00  None     20
3  7876  ADAMS    CLERK  7788  1987-07-13  1100.00  None     20
4  7900  JAMES    CLERK  7698  1981-12-03   950.00  None     30

____ 는 글자수 만큼 언더바를 써주면 된다. 
글자수가 4개인 스미스와 스캇을 출력해보자면

In [137]:
sql_25 = """

            SELECT * FROM emp WHERE ENAME LIKE 'S____'
        """
result_24 = _dt.executeAll(sql_25)
pd.DataFrame(result_24)

EMPNO  ENAME      JOB   MGR    HIREDATE      SAL  COMM DEPTNO
0  7369  SMITH    CLERK  7902  1980-12-17   800.00  None     20
1  7788  SCOTT  ANALYST  7566  1987-06-28  3000.00  None     20

In [149]:
sql_25 = """

            SELECT * FROM emp WHERE ENAME LIKE '%%E_'
        """
result_24 = _dt.executeAll(sql_25)
pd.DataFrame(result_24)

EMPNO   ENAME       JOB   MGR    HIREDATE      SAL    COMM DEPTNO
0  7499   ALLEN  SALESMAN  7698  1981-02-20  1600.00  300.00     30
1  7566   JONES   MANAGER  7839  1981-04-02  2975.00    None     20
2  7844  TURNER  SALESMAN  7698  1981-09-08  1500.00    0.00     30
3  7900   JAMES     CLERK  7698  1981-12-03   950.00    None     30
4  7934  MILLER     CLERK  7782  1982-01-23  1300.00    None     10

In [142]:
sql_25 = """

            SELECT * FROM emp WHERE ENAME LIKE '%%ES'
        """
result_24 = _dt.executeAll(sql_25)
pd.DataFrame(result_24)

EMPNO  ENAME      JOB   MGR    HIREDATE      SAL  COMM DEPTNO
0  7566  JONES  MANAGER  7839  1981-04-02  2975.00  None     20
1  7900  JAMES    CLERK  7698  1981-12-03   950.00  None     30

이렇게 조건을 하나씩 좁혀가면서 원하는 글자를 찾아낼 수 있다. 

실생활에서 어떨 때 like를 쓸 까? 
1. 고객정보가 있는데 김씨 성만 뽑고싶다. 이씨 성만 뽑고싶다. 
2. 특정 지역 주소만 뽑고 싶을때 서울특별시 강북구만 뽑고싶다면 쓸 수 있다. 
3. 검색엔진에서 단어 하나 적으면 연관된 검색어가 다 나온다. 

등등 검색에 특화된 활용법이 나올 수 있다. 